examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+	{
+		LOG("Window window moved to { %d, %d }", x, y);
+	}
+	void onWindowResized_impl(uint32_t w, uint32_t h) override
+	{
+		LOG("Window resized to { %u, %u }", w, h);
+	}
+	void onWindowMinimized_impl() override
+	{
+		LOG("Window minimized");
+	}
+	void onWindowMaximized_impl() override
+	{
+		LOG("Window maximized");
+	}
+	void onGainedMouseFocus_impl() override
+	{
+		LOG("Window gained mouse focus");
+	}
+	void onLostMouseFocus_impl() override
+	{
+		LOG("Window lost mouse focus");
+	}
+	void onGainedKeyboardFocus_impl() override
+	{
+		LOG("Window gained keyboard focus");
+	}
+	void onLostKeyboardFocus_impl() override
+	{
+		LOG("Window lost keyboard focus");
+	}
+
+	void onMouseConnected_impl(core::smart_refctd_ptr<nbl::ui::IMouseEventChannel>&& mch) override
+	{
+		LOG("A mouse has been connected");
+	}
+	void onMouseDisconnected_impl(nbl::ui::IMouseEventChannel* mch) override
+	{
+		LOG("A mouse has been disconnected");
+	}
+	void onKeyboardConnected_impl(core::smart_refctd_ptr<nbl::ui::IKeyboardEventChannel>&& kbch) override
+	{
+		LOG("A keyboard has been connected");
+	}
+	void onKeyboardDisconnected_impl(nbl::ui::IKeyboardEventChannel* mch) override
+	{
+		LOG("A keyboard has been disconnected");
+	}
+};
 
 int main()
 {
 	constexpr uint32_t WIN_W = 800u;
 	constexpr uint32_t WIN_H = 600u;
 	constexpr uint32_t SC_IMG_COUNT = 3u;
-
-	auto win = CWindowT::create(WIN_W, WIN_H, system::IWindow::ECF_NONE);
-
+	
+	// Note(achal): This is unused, for now
 	video::SDebugCallback dbgcb;
 	dbgcb.callback = &debugCallback;
 	dbgcb.userData = nullptr;
-	auto gl = video::IAPIConnection::create(video::EAT_OPENGL, 0, "New API Test", &dbgcb);
-	auto surface = gl->createSurface(win.get());
+	
+	auto system = CommonAPI::createSystem();
+
+	auto winManager = core::make_smart_refctd_ptr<nbl::ui::CWindowManagerWin32>();
+
+	nbl::ui::IWindow::SCreationParams params;
+	params.callback = nullptr;
+	params.width = WIN_W;
+	params.height = WIN_H;
+	params.x = 0;
+	params.y = 0;
+	params.system = core::smart_refctd_ptr(system);
+	params.flags = nbl::ui::IWindow::ECF_NONE;
+	params.windowCaption = "Test Window";
+	params.callback = core::make_smart_refctd_ptr<DemoEventCallback>();
+	auto window = winManager->createWindow(std::move(params));
+
+	core::smart_refctd_ptr<video::IAPIConnection> vk = video::IAPIConnection::create(std::move(system), video::EAT_VULKAN, 0, "New API Test", &dbgcb);
+	core::smart_refctd_ptr<video::ISurface> surface = vk->createSurface(window.get());
+
+	auto gpus = vk->getPhysicalDevices();
+	assert(!gpus.empty());
+	
+	// Find a suitable gpu
+	// 
+	// Todo(achal): This process gets quite involved in Vulkan, do we want to delegate some of it to the engine.
+	// For example, the user just specifies if the application is a headless compute one and we could check the
+	// availability of the required extensions on the backend and report which physical device is suitable or
+	// some thing like that

```

Comment:
in later examples, deal with it via `CommonAPI.h`

Reply:
You could show the future users how to "manually" create a graphics + presenting application in example 01
then how to "manually" create a headless compute device in some other examples (i.e. in example 02 you could showcase compute + present)

Then in CommonAPI have a static method `choosePhysicalDeviceWithSingleQueueFamily<bool graphics, bool present>` that would return the gpu and queue family index